pkg/dcmtime/common.go

```diff
@@ -0,0 +1,147 @@
+package dcmtime
+
+import (
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+)
+
+// isIncluded returns whether `check` is included in `limit`.
+//
+// Example: to test whether seconds should be included, you would:
+// isIncluded(PrecisionSeconds, [caller-passed-limit])
+func isIncluded(check PrecisionLevel, precision PrecisionLevel) bool {
+	return check <= precision
+}
+
+// truncateMilliseconds truncate nanosecond time.Time value to arbitrary precision.
+func truncateMilliseconds(
+	nanoSeconds int, precision PrecisionLevel,
+) (millisecondsStr string) {
+	milliseconds := nanoSeconds / 1000
+	millisecondsStr = fmt.Sprintf("%06d", milliseconds)
+	millisecondsStr = millisecondsStr[:6-(PrecisionFull-precision)]
+
+	return millisecondsStr
+}
+
+// Const for time.Time timezone. We don't want a timezone because we don't REALLY know
+// that a TM or DA value is UTC, so we don't want UTC to be made explicit.
+var zeroTimezone = time.FixedZone("", 0)
+
+// durationInfo holds parsed piece of DA, TM, or DT info.
+type durationInfo struct {
+	// The parsed int value
+	Value int
+	// Whether it was present in the string
+	Present bool
+	// Only used when parsing Fractal seconds: the level of precision used.
+	FractalPrecision PrecisionLevel
+}
+
+// extractDurationInfo extracts a piece of DA, TM, or DT info from a parsed regex.
+func extractDurationInfo(subMatches []string, index int, fractal bool) (
+	info durationInfo, err error,
+) {
+	if len(subMatches) <= index {
+		return info, errors.New("not enough sub-matches")
+	}
+
+	valueStr := subMatches[index]
+	// If there was no match for the specific subgroup, this value is 0.
+	if valueStr == "" {
+		valueStr = "0"
+	} else {
+		info.Present = true
+	}
+
+	// If this is a fractal seconds value, we need to pad it out to nanoseconds for go.
+	if fractal {
+		// The fractal value can be any length, with truncation implying a loss of
+		// precision, we need to add trailing zeros to the hundred-millionth place to
+		// get our nano-seconds.
+		missingPlaces := 9 - len(valueStr)
+		valueStr = valueStr + strings.Repeat("0", missingPlaces)
+		info.FractalPrecision = PrecisionFull - PrecisionLevel(missingPlaces-3)
+	}
+
+	info.Value, err = strconv.Atoi(valueStr)
+	if err != nil {
+		return info, fmt.Errorf("error parsing int: %w", err)
+	}
+
+	return info, nil
+}
+
+// updatePrecision takes in our current precision and a piece of parsed info, then
+// updates the precision based on whether the info was present.
+//
+// levelIsFull should be set to true if the level we are checking is "Full" for the
+// type of value we are parsing.
+func updatePrecision(
+	current PrecisionLevel,
+	info durationInfo,
+	infoLevel PrecisionLevel,
+	levelIsFull bool,
```

Comment:
You can group the two precision levels together (`func updatePrecision(... current, info PrecisionLevel)`)